File: src/runtime2/js/pytra/png.js
// source: src/pytra/utils/png.py
// PNG 画像出力ヘルパ（JavaScript版）。

const fs = require("node:fs");
const path = require("node:path");

function u16le(v) {
  return Buffer.from([v & 0xff, (v >>> 8) & 0xff]);
}

function u32be(v) {
  return Buffer.from([(v >>> 24) & 0xff, (v >>> 16) & 0xff, (v >>> 8) & 0xff, v & 0xff]);
}

function crc32(buf) {
  let crc = 0xffffffff;
  for (let i = 0; i < buf.length; i += 1) {
    crc ^= buf[i];
    for (let j = 0; j < 8; j += 1) {
      if ((crc & 1) !== 0) {
        crc = (crc >>> 1) ^ 0xedb88320;
      } else {
        crc >>>= 1;
      }
    }
  }
  return (~crc) >>> 0;
}

function adler32(buf) {
  const mod = 65521;
  let s1 = 1;
  let s2 = 0;
  for (let i = 0; i < buf.length; i += 1) {
    s1 += buf[i];
    if (s1 >= mod) {
      s1 -= mod;
    }
    s2 += s1;
    s2 %= mod;
  }
  return (((s2 << 16) >>> 0) | (s1 & 0xffff)) >>> 0;
}

function zlibDeflateStore(data) {
  const out = [];
  out.push(Buffer.from([0x78, 0x01]));
  let pos = 0;
  while (pos < data.length) {
    const remain = data.length - pos;
    const chunkLen = remain > 65535 ? 65535 : remain;
    const final = pos + chunkLen >= data.length ? 1 : 0;
    out.push(Buffer.from([final]));
    out.push(u16le(chunkLen));
    out.push(u16le((0xffff ^ chunkLen) & 0xffff));
    out.push(data.subarray(pos, pos + chunkLen));
    pos += chunkLen;
  }
  out.push(u32be(adler32(data)));
  return Buffer.concat(out);
}

function chunk(chunkType, data) {
  const crcInput = Buffer.concat([Buffer.from(chunkType, "ascii"), data]);
  return Buffer.concat([u32be(data.length >>> 0), Buffer.from(chunkType, "ascii"), data, u32be(crc32(crcInput))]);
}

function write_rgb_png(outPath, width, height, pixels) {
  const raw = Buffer.from(pixels);
  const expected = width * height * 3;
  if (raw.length !== expected) {
    throw new Error(`pixels length mismatch: got=${raw.length} expected=${expected}`);
  }

  const rowBytes = width * 3;
  const scanlines = Buffer.alloc(height * (rowBytes + 1));
  for (let y = 0; y < height; y += 1) {
    const dst = y * (rowBytes + 1);
    scanlines[dst] = 0;
    raw.copy(scanlines, dst + 1, y * rowBytes, (y + 1) * rowBytes);
  }

  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8;
  ihdr[9] = 2;
  ihdr[10] = 0;
  ihdr[11] = 0;
  ihdr[12] = 0;

  const idat = zlibDeflateStore(scanlines);
  const png = Buffer.concat([
    Buffer.from([0x89, 0x50, 0x4e, 0x47, 0x0d, 0x0a, 0x1a, 0x0a]),
    chunk("IHDR", ihdr),
    chunk("IDAT", idat),
    chunk("IEND", Buffer.alloc(0)),
  ]);

  fs.mkdirSync(path.dirname(outPath), { recursive: true });
  fs.writeFileSync(outPath, png);
}

module.exports = { write_rgb_png };

File: src/pytra/utils/png.py
"""PNG 書き出しユーティリティ（Python実行用）。

このモジュールは sample/py のスクリプトから利用し、
RGB 8bit バッファを PNG ファイルとして保存する。
"""

from __future__ import annotations


def _png_append_list(dst: list[int], src: list[int]) -> None:
    i = 0
    n = len(src)
    while i < n:
        dst.append(src[i])
        i += 1


def _crc32(data: list[int]) -> int:
    crc = 0xFFFFFFFF
    poly = 0xEDB88320
    for b in data:
        crc = crc ^ b
        i = 0
        while i < 8:
            lowbit = crc & 1
            if lowbit != 0:
                crc = (crc >> 1) ^ poly
            else:
                crc = crc >> 1
            i += 1
    return crc ^ 0xFFFFFFFF


def _adler32(data: list[int]) -> int:
    mod = 65521
    s1 = 1
    s2 = 0
    for b in data:
        s1 += b
        if s1 >= mod:
            s1 -= mod
        s2 += s1
        s2 = s2 % mod
    return ((s2 << 16) | s1) & 0xFFFFFFFF


def _png_u16le(v: int) -> list[int]:
    return [v & 0xFF, (v >> 8) & 0xFF]


def _png_u32be(v: int) -> list[int]:
    return [(v >> 24) & 0xFF, (v >> 16) & 0xFF, (v >> 8) & 0xFF, v & 0xFF]


def _zlib_deflate_store(data: list[int]) -> list[int]:
    out: list[int] = []
    _png_append_list(out, [0x78, 0x01])
    n = len(data)
    pos = 0
    while pos < n:
        remain = n - pos
        chunk_len = 65535 if remain > 65535 else remain
        final = 1 if (pos + chunk_len) >= n else 0
        out.append(final)
        _png_append_list(out, _png_u16le(chunk_len))
        _png_append_list(out, _png_u16le(0xFFFF ^ chunk_len))
        i = pos
        end = pos + chunk_len
        while i < end:
            out.append(data[i])
            i += 1
        pos += chunk_len
    _png_append_list(out, _png_u32be(_adler32(data)))
    return out


def _chunk(chunk_type: list[int], data: list[int]) -> list[int]:
    crc_input: list[int] = []
    _png_append_list(crc_input, chunk_type)
    _png_append_list(crc_input, data)
    crc = _crc32(crc_input) & 0xFFFFFFFF
    out: list[int] = []
    _png_append_list(out, _png_u32be(len(data)))
    _png_append_list(out, chunk_type)
    _png_append_list(out, data)
    _png_append_list(out, _png_u32be(crc))
    return out


def write_rgb_png(path: str, width: int, height: int, pixels: bytes) -> None:
    raw: list[int] = []
    for b in pixels:
        raw.append(int(b))
    expected = width * height * 3
    if len(raw) != expected:
        raise ValueError("pixels length mismatch: got=" + str(len(raw)) + " expected=" + str(expected))

    scanlines: list[int] = []
    row_bytes = width * 3
    y = 0
    while y < height:
        scanlines.append(0)
        start = y * row_bytes
        end = start + row_bytes
        i = start
        while i < end:
            scanlines.append(raw[i])
            i += 1
        y += 1

    ihdr: list[int] = []
    _png_append_list(ihdr, _png_u32be(width))
    _png_append_list(ihdr, _png_u32be(height))
    _png_append_list(ihdr, [8, 2, 0, 0, 0])
    idat = _zlib_deflate_store(scanlines)

    png: list[int] = []
    _png_append_list(png, [137, 80, 78, 71, 13, 10, 26, 10])
    _png_append_list(png, _chunk([73, 72, 68, 82], ihdr))
    _png_append_list(png, _chunk([73, 68, 65, 84], idat))
    iend_data: list[int] = []
    _png_append_list(png, _chunk([73, 69, 78, 68], iend_data))

    with open(path, "wb") as f:
        f.write(bytes(png))
